externals/llvm-external-projects/torch-mlir-dialects/lib/Conversion/TcpToLinalg/Elementwise.cpp

```diff
@@ -0,0 +1,120 @@
+//===------------------------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// Also available under a BSD-style license. See LICENSE.
+//
+//===----------------------------------------------------------------------===//
+
+#include "torch-mlir-dialects/Conversion/TcpToLinalg/TcpToLinalg.h"
+
+#include "../PassDetail.h"
+#include "PopulatePatterns.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "torch-mlir-dialects/Dialect/Tcp/IR/TcpDialect.h"
+#include "torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.h"
+
+using namespace mlir;
+using namespace mlir::tcp;
+
+namespace {
+
+Value createElementwiseLinalgGeneric(
+    OpBuilder &b, Location loc, ValueRange tensorOperands,
+    RankedTensorType resultTensorType,
+    function_ref<void(OpBuilder&, Location, ValueRange)> bodyBuilder) {
+  auto resultRank = resultTensorType.getRank();
+  SmallVector<Value> resultDimSizes;
+  SmallVector<AffineMap> indexingMaps;
+
+  // In order to populate the resultDimSizes, we only need to look at one of
+  // the tensorOperands, since all the operands are expected to have the same
+  // shape.
+  auto tensorOperand = tensorOperands[0];
+  for (int64_t i = 0; i < resultRank; ++i) {
+    resultDimSizes.push_back(b.createOrFold<tensor::DimOp>(loc, tensorOperand, i));
+  }
+
+  // Add indexing maps for all the tensor operands and for the result.
+  for (size_t i = 0; i < tensorOperands.size(); ++i) {
+    indexingMaps.push_back(b.getMultiDimIdentityMap(resultRank));
+  }
+  indexingMaps.push_back(b.getMultiDimIdentityMap(resultRank));
+
+  SmallVector<StringRef> iteratorTypes(resultRank,
+                                       getParallelIteratorTypeName());
+
+  Value emptyTensor = b.create<tensor::EmptyOp>(
+      loc, getAsOpFoldResult(resultDimSizes), resultTensorType.getElementType());
+  return b.create<linalg::GenericOp>(
+      loc, emptyTensor.getType(), tensorOperands, emptyTensor, indexingMaps,
+      iteratorTypes, bodyBuilder).getResult(0);
+}
+
+template<typename TcpOpT>
+Value createLinalgPayloadForElementwiseOp(
```

Comment:
I am not entirely sure why this is templated. It would probably be the same if you use `Operation *op` instead of `TcpOpT op`?

Reply:
Maybe also make this return `FailureOr<Value>`. Not a fan of returning `nullptr`s around.